src/ruby/qps/proxy-worker.rb

```diff
@@ -87,18 +87,24 @@ def cpu_cores
   # status return value.
   def run_client(reqs)
     q = EnumeratorQueue.new(self)
+    mark_ind = 0
     Thread.new {
       reqs.each do |req|
         case req.argtype.to_s
         when 'setup'
           @bmc.setup(req.setup)
           q.push(Grpc::Testing::ClientStatus.new(stats: @bmc.mark(false)))
         when 'mark'
+          # There are 3 marks: the first is for initiating the hist;
+          # The second is for warm-up; The last is for the benchmark.
```

Comment:
Hi there,
It shouldn't be assumed that there are only 3 marks. It is possible that we would want more statistics reporting points than that, and all the benchmarks are written to assume that there can be an unlimited number of marks. That's why the mark has a reset field; to allow some marks to specify that the stats should be cleared. It's fine to react to marks differently from other operations, but I don't think you should assume a count of marks.

Reply:
Thanks for the correction. Then I also can't assume that only the last mark is always for the benchmark. 
If there is a middle "mark" used for the benchmark, uploading the data still consumes time for the next mark, which lower the qps. I didn't see where I can pause the drive, letting ruby finish uploading before start next "mark".
I guess I can continue the signal way to let php know when "mark" is called. But instead of uploading all latencies, I should implement histogram in the php side and upload it once before each mark. \
Do you have some suggestions to let php knows when to upload other than using signal? The response from the server seems useless because it remains the same until it is closed.
n the benchmark framework I

Reply:
Is that right that "reqs.each do |req|" works as 
while (1){
    in = stream.recv(); // the program will hang there for warm-up time, lots of benchmark time.
    case in.isSetup:
    case in.inMark: get data from php, upload histogram to the driver.
}
Thus in the in.inMark, streamming upload will always be expensive and it cost time from the next "mark". What I'm doing now is to make the split latency sent together, which may speed up cache or buffer re-use or something. But the best way is just uploading once, right?

Reply:
I think the right thing is to restructure the proxy worker. Right now the PHP client talks to the proxy worker over a 1-sided client-side streaming RPC called ReportTime. Instead, it should be changed to a bidi-streaming RPC. On the bidi streaming RPC, the proxy worker should send each mark to the PHP client. The PHP client should read that and then respond with a histogram. The proxy worker can then pass that histogram back to the driver. That will require maintaining a histogram in the PHP client and sending a ClientStats proto back to the proxy worker on a mark. Does that make sense? Maybe the Quit and other messages could also be passed from the proxy worker to the PHP client like that to take signal out of the mix.

Reply:
That sounds cool by using a bidi-streaming. I'll try use async way or fork a child process listening to this bidi-streaming RPC in sync way. Thanks for your help!
I'll close this PR and reopen it until I finish them.

Reply:
Oh yes, I forgot about that point, that neither of those options is really natural in PHP. I think that's why I went with the 1-directional streaming thing in the first place. Probably needs better investigation then.

Reply:
Another possibility: go with the example that you're currently using (storing the latency results in an array and sending it out as a batch) but send it out periodically like once every 100 ms or something so that you get the batching effects that you're talking about but without having to alter the framework altogether.

Reply:
I'm still using signal to let them talk to each other. I tried create a child process and let a streaming RPC listening there. But it seems php-gRPC doesn't work on child process(the same code works in the parent, will hang in the child). I don't know it's php's problem or php-gRPC's problem yet.
Currently every mark will trigger php uploading, and continue until the data is uploaded. 
I'll implement histogram in the php side next.